DeviceAdministration/Infrastructure.UnitTests/Infrastructure/DeviceRegistryRepositoryTests.cs

```diff
@@ -0,0 +1,252 @@
+    {
+        private readonly Fixture _fixture;
+        private readonly DeviceRegistryRepository _deviceRegistryRepository;
+        private readonly Mock<IDocumentDBClient<DeviceModel>> _mockDocumentDBClient;
+
+        public DeviceRegistryRepositoryTests()
+        {
+            this._fixture = new Fixture();
+            this._mockDocumentDBClient = new Mock<IDocumentDBClient<DeviceModel>>();
+            this._deviceRegistryRepository = new DeviceRegistryRepository(this._mockDocumentDBClient.Object);
+        }
+
+       
+        [Theory]
+        [InlineData(null)]
+        [InlineData("")]
+        public async Task GetDeviceAsync_should_throw_argument_exception_when_device_id_is_invalid(string deviceId)
+        {
+            await Assert.ThrowsAsync<ArgumentException>(() => this._deviceRegistryRepository.GetDeviceAsync(deviceId));
+        }
+
+        [Fact]
+        public async Task GetDeviceAsync_should_get_device_from_client()
+        {
+            var devices = this._fixture.Create<List<DeviceModel>>();
+            var deviceId = devices.First().DeviceProperties.DeviceID;
+
+            this._mockDocumentDBClient
+                .Setup(x => x.QueryAsync())
+                .ReturnsAsync(devices.AsQueryable());
+
+            var device = await this._deviceRegistryRepository.GetDeviceAsync(deviceId);
+
+            Assert.NotNull(device);
+            Assert.Same(devices[0], device);
+        }
+
+        [Fact]
+        public async Task GetDeviceAsync_should_return_null_when_device_id_is_not_found()
+        {
+            var devices = this._fixture.Create<List<DeviceModel>>();
+
+            this._mockDocumentDBClient
+                .Setup(x => x.QueryAsync())
+                .ReturnsAsync(devices.AsQueryable());
+
+            var device = await this._deviceRegistryRepository.GetDeviceAsync("foobarbaz");
+
+            Assert.Null(device);
+        }
+
+        [Fact]
+        public async Task AddDeviceAsync_should_throw_argument_null_when_device_is_null()
+        {
+            await Assert.ThrowsAsync<ArgumentNullException>(() => this._deviceRegistryRepository.AddDeviceAsync(null));
+        }
+
+        [Fact]
+        public async Task AddDeviceAsync_should_throw_device_exists_exception_when_device_with_same_id_exists()
+        {
+            var devices = this._fixture.Create<List<DeviceModel>>();
+            this._mockDocumentDBClient
+                .Setup(x => x.QueryAsync())
+                .ReturnsAsync(devices.AsQueryable());
+
+            await Assert.ThrowsAsync<DeviceAlreadyRegisteredException>(() => this._deviceRegistryRepository.AddDeviceAsync(devices.First()));
+        }
+
+        [Fact]
+        public async Task AddDeviceAsync_should_add_device_to_db()
+        {
+            var device = this._fixture.Create<DeviceModel>();
+            await this._deviceRegistryRepository.AddDeviceAsync(device);
+            this._mockDocumentDBClient.Verify(x => x.SaveAsync(device));
```

Comment:
I don't have to. Mocks do auto call counting so mocking is not explicitly concept.